

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Import / ETL &mdash; GenomicsDB  documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Common Issues" href="common-issues.html" />
    <link rel="prev" title="Building + Installing" href="building-installing.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> GenomicsDB
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">GenomicsDB Docs</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview (DOCUMENTATION WORK-IN-PROGRESS)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="overview.html#concepts-terminology">Concepts + Terminology</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="storage.html">Supported Storage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="storage.html#local-cluster">Local / Cluster</a></li>
<li class="toctree-l2"><a class="reference internal" href="storage.html#cloud">Cloud</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="building-installing.html">Building + Installing</a><ul>
<li class="toctree-l2"><a class="reference internal" href="building-installing.html#bash">Bash</a></li>
<li class="toctree-l2"><a class="reference internal" href="building-installing.html#docker">Docker</a></li>
<li class="toctree-l2"><a class="reference internal" href="building-installing.html#cli-tools">CLI Tools</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Import / ETL</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#vcf-gvcf">VCF/gVCF</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#preliminaries">Preliminaries</a></li>
<li class="toctree-l3"><a class="reference internal" href="#organizing-your-data">Organizing your data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#csv">CSV</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#id2">Preliminaries</a></li>
<li class="toctree-l3"><a class="reference internal" href="#csv-format">CSV format</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#example">Example</a></li>
<li class="toctree-l4"><a class="reference internal" href="#special-fields">Special fields</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="#id5">Organizing your data</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#best-practices">Best Practices</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="common-issues.html">Common Issues</a></li>
</ul>
<p class="caption"><span class="caption-text">API Docs</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="apis.html">APIs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="apis.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="apis.html#c">C++</a></li>
<li class="toctree-l2"><a class="reference internal" href="apis.html#python">Python</a><ul>
<li class="toctree-l3"><a class="reference internal" href="apis.html#api-reference">API Reference</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="apis.html#java">Java</a><ul>
<li class="toctree-l3"><a class="reference internal" href="apis.html#id1">API Reference</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="apis.html#r">R</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="cli-tools.html">CLI Tools</a></li>
</ul>
<p class="caption"><span class="caption-text">Examples</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="examples/notebooks.html">Notebooks</a><ul>
<li class="toctree-l2"><a class="reference internal" href="examples/notebooks.html#python-notebook">Python Notebook</a></li>
<li class="toctree-l2"><a class="reference internal" href="examples/notebooks.html#scala-notebook">Scala Notebook</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="examples/import-examples.html">Import / ETL Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples/gatk.html">Using with GATK</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">GenomicsDB</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Import / ETL</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/import-etl.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="import-etl">
<span id="id1"></span><h1>Import / ETL<a class="headerlink" href="#import-etl" title="Permalink to this headline">¶</a></h1>
<p>GenomicsDB supports importing genomics data in several common formats, and with a variety of methods.</p>
<p>GenomicsDB can ingest data in VCF, gVCF. and CSV formats. When importing VCFs or gVCFs,
you need to run a few preprocessing steps before importing.</p>
<p>The primary and suggested method of importing is to use the native vcf2genomicsdb importer tool (see CLI Tools).
There is also a Java API for importing (see here).</p>
<div class="section" id="vcf-gvcf">
<h2>VCF/gVCF<a class="headerlink" href="#vcf-gvcf" title="Permalink to this headline">¶</a></h2>
<p>The import program can handle block compressed and indexed VCFs, gVCFs, BCFs and gBCFs.
For brevity, we will only use the term VCF.</p>
<div class="section" id="preliminaries">
<h3>Preliminaries<a class="headerlink" href="#preliminaries" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>You need libcsv while <a class="reference external" href="https://github.com/GenomicsDB/GenomicsDB/wiki/Compiling-GenomicsDB">compiling</a>.</p></li>
</ul>
<ul class="simple">
<li><p>License terms: We use libcsv to parse CSV files.
<a class="reference external" href="https://sourceforge.net/projects/libcsv/">libcsv</a> is licensed under the GNU Library or Lesser General Public License version 2.0 (LGPLv2).
So, if you are re-distributing binaries or object files, they may be subject to LGPLv2 terms.
Please ensure that any binaries/object files you distribute are compliant with LGPLv2.
You can disable libcsv usage by not setting the USE_LIBCSV and LIBCSV_DIR flags during compilation.
However, your binaries/executables will not be able to import CSV files into GenomicsDB.</p></li>
</ul>
</div>
<div class="section" id="organizing-your-data">
<h3>Organizing your data<a class="headerlink" href="#organizing-your-data" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>All your VCFs must be block compressed and indexed. <a class="reference external" href="https://github.com/samtools/bcftools">Bcftools</a> is one good option for compressing and indexing.</p></li>
</ul>
<ul class="simple">
<li><p>The VCF format allows you to have multiple lines with the same position (identical chromosome+pos). The import program can NOT handle such VCFs. Make sure you use tools like bcftools to collapse multiple lines with the same position into a single line. Example commands are <a class="reference external" href="https://github.com/GenomicsDB/GenomicsDB/wiki/Useful-external-tools#collapse-multiple-lines-with-the-same-genomic-position-into-a-single-line">provided here</a>.</p></li>
</ul>
<ul>
<li><dl>
<dt>In a multi-node environment, you must decide:</dt><dd><ul class="simple">
<li><p>How to <a class="reference external" href="https://github.com/GenomicsDB/GenomicsDB/wiki/GenomicsDB-setup-in-a-multi-node-cluster">partition your data in GenomicsDB</a>.</p></li>
</ul>
<ul class="simple">
<li><dl class="simple">
<dt>How your VCF files are accessed by the import program:</dt><dd><ul>
<li><p>On a shared filesystem (NFS, Lustre etc) accessible from all nodes.</p></li>
<li><p>If you are partitioning data by rows, your files can be scattered across local filesystems on multiple machines; each filesystem accessible only by the node on which it is mounted. Currently, such a setup isn’t supported for column partitioning.</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</li>
</ul>
</div>
</div>
<div class="section" id="csv">
<h2>CSV<a class="headerlink" href="#csv" title="Permalink to this headline">¶</a></h2>
<p>Importing CSVs into GenomicsDB</p>
<div class="section" id="id2">
<h3>Preliminaries<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>You need libcsv while <a class="reference external" href="https://github.com/GenomicsDB/GenomicsDB/wiki/Compiling-GenomicsDB">compiling</a>.</p></li>
</ul>
<ul class="simple">
<li><p>License terms: We use libcsv to parse CSV files.
<a class="reference external" href="https://sourceforge.net/projects/libcsv/">libcsv</a> is licensed under the GNU Library or Lesser General Public License version 2.0 (LGPLv2).
So, if you are re-distributing binaries or object files, they may be subject to LGPLv2 terms.
Please ensure that any binaries/object files you distribute are compliant with LGPLv2.
You can disable libcsv usage by not setting the USE_LIBCSV and LIBCSV_DIR flags during compilation.
However, your binaries/executables will not be able to import CSV files into GenomicsDB.</p></li>
</ul>
</div>
<div class="section" id="csv-format">
<h3>CSV format<a class="headerlink" href="#csv-format" title="Permalink to this headline">¶</a></h3>
<p>Given a variant call at a specific position in the genome for a particular CallSet/sample,
the CSV file format supported by GenomicsDB contains one line describing the call.
Essentially, each line in the CSV describes one cell in the GenomicsDB array.</p>
<p>The exact format of the CSV is shown below:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;</span><span class="n">row</span><span class="o">&gt;</span><span class="p">,</span><span class="o">&lt;</span><span class="n">begin_column</span><span class="o">&gt;</span><span class="p">,</span><span class="o">&lt;</span><span class="n">end_column</span><span class="o">&gt;</span><span class="p">,</span><span class="o">&lt;</span><span class="n">REF</span><span class="o">&gt;</span><span class="p">,</span><span class="o">&lt;</span><span class="n">concatenated_ALT</span><span class="o">&gt;</span><span class="p">,</span><span class="o">&lt;</span><span class="n">QUAL</span><span class="o">&gt;</span><span class="p">,</span><span class="o">&lt;</span><span class="n">FILTER_field_specification</span><span class="o">&gt;</span><span class="p">[,</span><span class="o">&lt;</span><span class="n">other_fields_specification</span><span class="o">&gt;</span><span class="p">]</span>
</pre></div>
</div>
<p>Fixed fields:</p>
<ul class="simple">
<li><p>row (mandatory, type: int64): Row number in GenomicsDB for this sample/CallSet</p></li>
<li><p>begin_column (mandatory, type: int64): Column in GenomicsDB at which the variant call begins</p></li>
<li><p>end_column (mandatory, type: int64): Column in GenomicsDB at which the variant call ends (inclusive).</p></li>
<li><p>REF (mandatory, type:string): Reference allele at the given position.</p></li>
<li><p>concatenated_ALT (mandatory, type: string): Concatenated alternate alleles separated by the character ‘|’. For example, if a given call has two alternate alleles TAG and TG, then the concatenated string would be TAG|TG.</p></li>
<li><p>QUAL (optional, type: float): Represents the QUAL field in a VCF. It can be left empty implying that the value was missing for this call.</p></li>
<li><p>FILTER_field_specification (mandatory, type: variable length list of integers): Each element of the FILTER field list is an integer representing the FILTERs that this call failed (similar to how a BCF represents FILTERs). The first element of this field is an integer displaying the number of elements in the list. A value of 0 indicates that the list is empty.</p></li>
</ul>
<p>Additional fields can be optionally specified</p>
<ul>
<li><p>&lt;other_fields_specification&gt;: The format depends on the type of field:</p>
<blockquote>
<div><ul class="simple">
<li><p>String type fields (fixed or variable length strings): The field should contain the string - an empty token indicates that the field is missing.</p></li>
<li><p>Fixed length field (int or float): The field should contain exactly N elements where N is the length of the field (fixed constant). One or more tokens may be left empty to indicate that those elements are missing.</p></li>
<li><p>Variable length field (int or float): The first element of this field should be an integer denoting the number of elements in the field for this call. It should then be followed by the elements of this field. An empty or missing field can be specified by setting the first element (field length) to 0 - no other elements should follow an empty field.</p></li>
</ul>
</div></blockquote>
</li>
</ul>
<div class="section" id="example">
<h4>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h4>
<p>The following line contains 2 fields in addition to the fixed fields:</p>
<ul>
<li><p>SB: Fixed length field of 4 integers</p></li>
<li><p>PL: Variable length field of integers</p>
<blockquote>
<div><div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="mi">2</span><span class="p">,</span><span class="mi">1857210</span><span class="p">,</span><span class="mi">1857210</span><span class="p">,</span><span class="n">G</span><span class="p">,</span><span class="n">A</span><span class="o">|</span><span class="n">T</span><span class="p">,</span><span class="mf">894.77</span><span class="p">,</span><span class="mi">0</span><span class="p">,,,,,</span><span class="mi">6</span><span class="p">,</span><span class="mi">923</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">599</span><span class="p">,</span><span class="mi">996</span><span class="p">,</span><span class="mi">701</span><span class="p">,</span><span class="mi">1697</span>
</pre></div>
</div>
</div></blockquote>
</li>
</ul>
<p>The line specifies the variant call for row id 2, beginning at column 1857210 and ending at 1857210.
The REF allele is ‘G’ and the call has 2 alternate alleles ‘A’ and ‘T’ (SNVs).
The QUAL value is 894.77 and there are no FILTERs specified (hence FILTER field length = 0).
The SB field is missing - denoted by the 4 empty tokens.
The PL field consists of 6 integers - the length appears first (since PL is a variable length field) followed by the elements [923,0,599,996,701,1697].</p>
</div>
<div class="section" id="special-fields">
<h4>Special fields<a class="headerlink" href="#special-fields" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p>GT is represented in the CSV as a variable length list of integers - each element of the list refers to the allele index (0 for reference allele, 1 for the first alternate allele and so on). The length of the list represents the ploidy of the sample/CallSet and must be specified in the CSV line (since GT is treated as a variable length list).</p></li>
</ul>
</div>
</div>
<div class="section" id="id5">
<h3>Organizing your data<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>All CSV files imported into a GenomicsDB array must respect the number and order of fields as defined in the <a class="reference external" href="https://github.com/GenomicsDB/GenomicsDB/wiki/Importing-VCF-data-into-GenomicsDB#fields-information">vid_mapping_file</a>.</p></li>
</ul>
<ul class="simple">
<li><p>The import program cannot handle CSV files where 2 lines have the same value of row and begin_column - this restriction is similar to that imposed on <a class="reference external" href="https://github.com/GenomicsDB/GenomicsDB/wiki/Importing-VCF-data-into-GenomicsDB#fields-information">loading VCFs</a>.</p></li>
</ul>
</div>
</div>
<div class="section" id="best-practices">
<h2>Best Practices<a class="headerlink" href="#best-practices" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>best practices</p></li>
</ul>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="common-issues.html" class="btn btn-neutral float-right" title="Common Issues" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="building-installing.html" class="btn btn-neutral float-left" title="Building + Installing" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, GenomicsDB.org

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>