

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Overview (DOCUMENTATION WORK-IN-PROGRESS) &mdash; GenomicsDB  documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="#" class="icon icon-home" alt="Documentation Home"> GenomicsDB
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li><a class="reference internal" href="#">Overview (DOCUMENTATION WORK-IN-PROGRESS)</a><ul>
<li><a class="reference internal" href="#concepts-terminology">Concepts + Terminology</a></li>
</ul>
</li>
<li><a class="reference internal" href="#building-installing">Building + Installing</a></li>
<li><a class="reference internal" href="#apis">APIs</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#c">C++</a></li>
<li><a class="reference internal" href="#python">Python</a><ul>
<li><a class="reference internal" href="#api-reference">API Reference</a></li>
</ul>
</li>
<li><a class="reference internal" href="#java">Java</a><ul>
<li><a class="reference internal" href="#id3">API Reference</a></li>
</ul>
</li>
<li><a class="reference internal" href="#r">R</a></li>
<li><a class="reference internal" href="#spark">Spark</a></li>
</ul>
</li>
<li><a class="reference internal" href="#cli-tools">CLI Tools</a></li>
<li><a class="reference internal" href="#import-etl">Import / ETL</a><ul>
<li><a class="reference internal" href="#vcf">VCF</a></li>
<li><a class="reference internal" href="#gvcf">gVCF</a></li>
<li><a class="reference internal" href="#csv">CSV</a><ul>
<li><a class="reference internal" href="#preliminaries">Preliminaries</a></li>
<li><a class="reference internal" href="#csv-format">CSV format</a><ul>
<li><a class="reference internal" href="#example">Example</a></li>
<li><a class="reference internal" href="#special-fields">Special fields</a></li>
</ul>
</li>
<li><a class="reference internal" href="#organizing-your-data">Organizing your data</a></li>
</ul>
</li>
<li><a class="reference internal" href="#compression-schemas">Compression Schemas</a></li>
<li><a class="reference internal" href="#best-practices">Best Practices</a></li>
</ul>
</li>
<li><a class="reference internal" href="#supported-storage">Supported Storage</a><ul>
<li><a class="reference internal" href="#local-cluster">Local / Cluster</a></li>
<li><a class="reference internal" href="#cloud">Cloud</a></li>
</ul>
</li>
<li><a class="reference internal" href="#examples-tutorials">Examples / Tutorials</a><ul>
<li><a class="reference internal" href="#python-notebook">Python Notebook</a></li>
<li><a class="reference internal" href="#scala-notebook">Scala Notebook</a></li>
</ul>
</li>
<li><a class="reference internal" href="#advanced">Advanced</a><ul>
<li><a class="reference internal" href="#performance-tuning">Performance Tuning</a></li>
<li><a class="reference internal" href="#detailed-explanations">Detailed Explanations</a></li>
</ul>
</li>
<li><a class="reference internal" href="#using-with-gatk">Using with GATK</a></li>
<li><a class="reference internal" href="#documentation-work-in-progress">Documentation WORK-IN-PROGRESS</a><ul>
<li><a class="reference internal" href="#supported-platforms-and-filesystems">Supported platforms and filesystems:</a><ul>
<li><a class="reference internal" href="#included-are">Included are</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</div>
            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">GenomicsDB</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="#" class="icon icon-home"></a> &raquo;</li>
        
      <li>Overview (DOCUMENTATION WORK-IN-PROGRESS)</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="overview-documentation-work-in-progress">
<h1>Overview (DOCUMENTATION WORK-IN-PROGRESS)<a class="headerlink" href="#overview-documentation-work-in-progress" title="Permalink to this headline">¶</a></h1>
<p>GenomicsDB is a highly performant scalable data storage written in C++ for importing, querying and transforming genomic
variant data.</p>
<div class="section" id="concepts-terminology">
<span id="id1"></span><h2>Concepts + Terminology<a class="headerlink" href="#concepts-terminology" title="Permalink to this headline">¶</a></h2>
<p>This section outlines basic GenomicsDB concepts and terminology used throughout this documentation.</p>
<ul>
<li><p><em>Cell</em>: Cell of an array (for example: A[i][j] in a 2-D array)</p></li>
<li><p><em>Workspace</em>: A directory on the file system under which multiple GenomicsDB arrays can be stored. The underlying array stores some metadata files under the workspace directory. All GenomicsDB import/create programs assume that if the specified workspace directory exists, then it’s a valid workspace directory. If the user points to an existing directory which doesn’t contain the workspace metadata, the create/import programs will exit with an exception. If the directory doesn’t exist, then the programs will initialize a new workspace correctly.</p></li>
<li><p><em>Array</em>: Name of a GenomicsDB array</p>
<blockquote>
<div><p>Given a workspace and array name, the framework will store its data in the directory <em>&lt;workspace&gt;/&lt;array&gt;</em>.</p>
</div></blockquote>
</li>
<li><p><em>Column-major</em> and <em>row-major</em> ordering: Denotes the order in which cells are stored on disk.</p>
<blockquote>
<div><ul class="simple">
<li><p>Column major storage implies that cells belonging to the same column are stored contiguously on disk (cells belonging to the same column but different rows are sorted by row id). When cells are stored in column major order queries such as “retrieve all cells belonging to genomic position <em>X</em>” are fast due to high spatial locality. However, queries such as “retrieve cells for sample <em>Y</em>” are relatively slow. For GenomicsDB, we expect the former type of queries to be more frequent and hence, by default all arrays are stored in column major order (even when partitioning by rows across GenomicsDB instances).</p></li>
<li><p>Row major ordering implies that cells belonging to the same row are stored contiguously on disk.</p></li>
</ul>
</div></blockquote>
</li>
<li><p><em>Bulk importing and incremental importing</em>: <em>Bulk importing</em> of data implies that all the data is loaded at once into GenomicsDB and the array is never modified later on. <em>Incremental importing</em> implies that the array can be modified by adding new samples/CallSet over time. GenomicsDB support both modes of operation - however, repeated incremental updates (for example, incrementally adding one sample/CallSet at a time, ~1000 times) may lead to reduced performance during querying. If you are interested in why this occurs, please read the section on writing arrays to understand the concept of fragments and how updates are performed. Ideally, users should perform incremental imports a few times (~10s of times) per array and each import should contain a large number of samples.</p></li>
</ul>
</div>
</div>
<div class="section" id="building-installing">
<span id="id2"></span><h1>Building + Installing<a class="headerlink" href="#building-installing" title="Permalink to this headline">¶</a></h1>
<p>Coming Soon</p>
</div>
<div class="section" id="apis">
<h1>APIs<a class="headerlink" href="#apis" title="Permalink to this headline">¶</a></h1>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>GenomicsDB has support/bindings for the following langauges.
&lt;explanation of query jsons&gt;</p>
</div>
<div class="section" id="c">
<h2>C++<a class="headerlink" href="#c" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>native interface</p></li>
<li><p>Querying CLI tools available - see <a class="reference internal" href="#vcf2genomicsdb"><span class="std std-ref">vcf2genomicsdb</span></a></p></li>
</ul>
</div>
<div class="section" id="python">
<h2>Python<a class="headerlink" href="#python" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Python bindings for GenomicsDB. Requires Python version <em>&gt;= 3.6</em></p></li>
<li><p>Built using Cython.</p></li>
<li><p>Currently, there is only support for querying (no importing).</p></li>
<li><p>Data can be returned in either a numpy matrix or a list of lists.</p></li>
</ul>
<p>The GenomicsDB python bindings are <a class="reference external" href="https://pypi.org/project/genomicsdb/">published on PyPI</a>.</p>
<p>Install <code class="docutils literal notranslate"><span class="pre">genomicsdb</span></code> with the following command:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">genomicsdb</span>
</pre></div>
</div>
<div class="section" id="api-reference">
<h3>API Reference<a class="headerlink" href="#api-reference" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Coming Soon. See the <a class="reference external" href="https://github.com/nalinigans/GenomicsDB-Python">bindings source code</a> for the time being.</p></li>
</ul>
</div>
</div>
<div class="section" id="java">
<h2>Java<a class="headerlink" href="#java" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>java - variant context &lt;link&gt;</p></li>
<li><p>Data can be returned in either a VariantContext object or &lt;other&gt;.</p></li>
<li><p>The GenomicsDB java bindings are published to <a class="reference external" href="https://mvnrepository.com/artifact/org.genomicsdb/genomicsdb">Maven Central</a> every release.</p></li>
</ul>
<ul class="simple">
<li><p>Add GenomicsDB to your maven project with the following snippet:</p></li>
</ul>
<div class="highlight-java notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;!--</span> <span class="n">https</span><span class="p">:</span><span class="c1">//mvnrepository.com/artifact/org.genomicsdb/genomicsdb --&gt;</span>
<span class="o">&lt;</span><span class="n">dependency</span><span class="o">&gt;</span>
   <span class="o">&lt;</span><span class="n">groupId</span><span class="o">&gt;</span><span class="n">org</span><span class="p">.</span><span class="na">genomicsdb</span><span class="o">&lt;/</span><span class="n">groupId</span><span class="o">&gt;</span>
   <span class="o">&lt;</span><span class="n">artifactId</span><span class="o">&gt;</span><span class="n">genomicsdb</span><span class="o">&lt;/</span><span class="n">artifactId</span><span class="o">&gt;</span>
   <span class="o">&lt;</span><span class="n">version</span><span class="o">&gt;</span><span class="mf">1.3.0</span><span class="o">&lt;/</span><span class="n">version</span><span class="o">&gt;</span>
<span class="o">&lt;/</span><span class="n">dependency</span><span class="o">&gt;</span>
</pre></div>
</div>
<div class="section" id="id3">
<h3>API Reference<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Coming Soon. See the <a class="reference external" href="https://github.com/GenomicsDB/GenomicsDB/tree/master/src/main/java/org/genomicsdb">source code</a> for the time being.</p></li>
</ul>
</div>
</div>
<div class="section" id="r">
<h2>R<a class="headerlink" href="#r" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Support Coming Soon!</p></li>
</ul>
</div>
<div class="section" id="spark">
<h2>Spark<a class="headerlink" href="#spark" title="Permalink to this headline">¶</a></h2>
</div>
</div>
<div class="section" id="cli-tools">
<span id="clis-tools"></span><h1>CLI Tools<a class="headerlink" href="#cli-tools" title="Permalink to this headline">¶</a></h1>
<p>GenomicsDB contains several command line tools. These executables are not published online,
you must <a class="reference internal" href="#building-installing"><span class="std std-ref">build and install</span></a> GenomicsDB to use them.</p>
<ul class="simple" id="vcf2genomicsdb">
<li><p>vcf2genomicsdb</p></li>
<li><p>gt mpi gather</p></li>
<li><p>create genomicsdb workspace</p></li>
<li><p>consolidate genomicsdb array</p></li>
<li><p>histogram</p></li>
<li><p>vcfdiff</p></li>
</ul>
</div>
<div class="section" id="import-etl">
<span id="id4"></span><h1>Import / ETL<a class="headerlink" href="#import-etl" title="Permalink to this headline">¶</a></h1>
<p>GenomicsDB supports importing genomics data in several common formats, and with a variety of methods.</p>
<p>GenomicsDB can ingest data in VCF, gVCF. and CSV formats. When importing VCFs or gVCFs,
you need to run a few preprocessing steps before importing.</p>
<p>The primary and suggested method of importing is to use the native vcf2genomicsdb importer tool (see CLI Tools).
There is also a Java API for importing (see here).</p>
<div class="section" id="vcf">
<h2>VCF<a class="headerlink" href="#vcf" title="Permalink to this headline">¶</a></h2>
<p>preprosessing steps</p>
</div>
<div class="section" id="gvcf">
<h2>gVCF<a class="headerlink" href="#gvcf" title="Permalink to this headline">¶</a></h2>
<p>preprosessing steps</p>
</div>
<div class="section" id="csv">
<h2>CSV<a class="headerlink" href="#csv" title="Permalink to this headline">¶</a></h2>
<p>Importing CSVs into GenomicsDB</p>
<div class="section" id="preliminaries">
<h3>Preliminaries<a class="headerlink" href="#preliminaries" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>You need libcsv while <a class="reference external" href="https://github.com/GenomicsDB/GenomicsDB/wiki/Compiling-GenomicsDB">compiling</a>.</p></li>
</ul>
<ul class="simple">
<li><p>License terms: We use libcsv to parse CSV files.
<a class="reference external" href="https://sourceforge.net/projects/libcsv/">libcsv</a> is licensed under the GNU Library or Lesser General Public License version 2.0 (LGPLv2).
So, if you are re-distributing binaries or object files, they may be subject to LGPLv2 terms.
Please ensure that any binaries/object files you distribute are compliant with LGPLv2.
You can disable libcsv usage by not setting the USE_LIBCSV and LIBCSV_DIR flags during compilation.
However, your binaries/executables will not be able to import CSV files into GenomicsDB.</p></li>
</ul>
</div>
<div class="section" id="csv-format">
<h3>CSV format<a class="headerlink" href="#csv-format" title="Permalink to this headline">¶</a></h3>
<p>Given a variant call at a specific position in the genome for a particular CallSet/sample,
the CSV file format supported by GenomicsDB contains one line describing the call.
Essentially, each line in the CSV describes one cell in the GenomicsDB array.</p>
<p>The exact format of the CSV is shown below:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="o">&lt;</span><span class="n">row</span><span class="o">&gt;</span><span class="p">,</span><span class="o">&lt;</span><span class="n">begin_column</span><span class="o">&gt;</span><span class="p">,</span><span class="o">&lt;</span><span class="n">end_column</span><span class="o">&gt;</span><span class="p">,</span><span class="o">&lt;</span><span class="n">REF</span><span class="o">&gt;</span><span class="p">,</span><span class="o">&lt;</span><span class="n">concatenated_ALT</span><span class="o">&gt;</span><span class="p">,</span><span class="o">&lt;</span><span class="n">QUAL</span><span class="o">&gt;</span><span class="p">,</span><span class="o">&lt;</span><span class="n">FILTER_field_specification</span><span class="o">&gt;</span><span class="p">[,</span><span class="o">&lt;</span><span class="n">other_fields_specification</span><span class="o">&gt;</span><span class="p">]</span>
</pre></div>
</div>
<p>Fixed fields:</p>
<ul class="simple">
<li><p>row (mandatory, type: int64): Row number in GenomicsDB for this sample/CallSet</p></li>
<li><p>begin_column (mandatory, type: int64): Column in GenomicsDB at which the variant call begins</p></li>
<li><p>end_column (mandatory, type: int64): Column in GenomicsDB at which the variant call ends (inclusive).</p></li>
<li><p>REF (mandatory, type:string): Reference allele at the given position.</p></li>
<li><p>concatenated_ALT (mandatory, type: string): Concatenated alternate alleles separated by the character ‘|’. For example, if a given call has two alternate alleles TAG and TG, then the concatenated string would be TAG|TG.</p></li>
<li><p>QUAL (optional, type: float): Represents the QUAL field in a VCF. It can be left empty implying that the value was missing for this call.</p></li>
<li><p>FILTER_field_specification (mandatory, type: variable length list of integers): Each element of the FILTER field list is an integer representing the FILTERs that this call failed (similar to how a BCF represents FILTERs). The first element of this field is an integer displaying the number of elements in the list. A value of 0 indicates that the list is empty.</p></li>
</ul>
<p>Additional fields can be optionally specified</p>
<ul>
<li><p>&lt;other_fields_specification&gt;: The format depends on the type of field:</p>
<blockquote>
<div><ul class="simple">
<li><p>String type fields (fixed or variable length strings): The field should contain the string - an empty token indicates that the field is missing.</p></li>
<li><p>Fixed length field (int or float): The field should contain exactly N elements where N is the length of the field (fixed constant). One or more tokens may be left empty to indicate that those elements are missing.</p></li>
<li><p>Variable length field (int or float): The first element of this field should be an integer denoting the number of elements in the field for this call. It should then be followed by the elements of this field. An empty or missing field can be specified by setting the first element (field length) to 0 - no other elements should follow an empty field.</p></li>
</ul>
</div></blockquote>
</li>
</ul>
<div class="section" id="example">
<h4>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h4>
<p>The following line contains 2 fields in addition to the fixed fields:</p>
<ul>
<li><p>SB: Fixed length field of 4 integers</p></li>
<li><p>PL: Variable length field of integers</p>
<blockquote>
<div><div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="mi">2</span><span class="p">,</span><span class="mi">1857210</span><span class="p">,</span><span class="mi">1857210</span><span class="p">,</span><span class="n">G</span><span class="p">,</span><span class="n">A</span><span class="o">|</span><span class="n">T</span><span class="p">,</span><span class="mf">894.77</span><span class="p">,</span><span class="mi">0</span><span class="p">,,,,,</span><span class="mi">6</span><span class="p">,</span><span class="mi">923</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">599</span><span class="p">,</span><span class="mi">996</span><span class="p">,</span><span class="mi">701</span><span class="p">,</span><span class="mi">1697</span>
</pre></div>
</div>
</div></blockquote>
</li>
</ul>
<p>The line specifies the variant call for row id 2, beginning at column 1857210 and ending at 1857210.
The REF allele is ‘G’ and the call has 2 alternate alleles ‘A’ and ‘T’ (SNVs).
The QUAL value is 894.77 and there are no FILTERs specified (hence FILTER field length = 0).
The SB field is missing - denoted by the 4 empty tokens.
The PL field consists of 6 integers - the length appears first (since PL is a variable length field) followed by the elements [923,0,599,996,701,1697].</p>
</div>
<div class="section" id="special-fields">
<h4>Special fields<a class="headerlink" href="#special-fields" title="Permalink to this headline">¶</a></h4>
<ul class="simple">
<li><p>GT is represented in the CSV as a variable length list of integers - each element of the list refers to the allele index (0 for reference allele, 1 for the first alternate allele and so on). The length of the list represents the ploidy of the sample/CallSet and must be specified in the CSV line (since GT is treated as a variable length list).</p></li>
</ul>
</div>
</div>
<div class="section" id="organizing-your-data">
<h3>Organizing your data<a class="headerlink" href="#organizing-your-data" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>All CSV files imported into a GenomicsDB array must respect the number and order of fields as defined in the <a class="reference external" href="https://github.com/GenomicsDB/GenomicsDB/wiki/Importing-VCF-data-into-GenomicsDB#fields-information">vid_mapping_file</a>.</p></li>
</ul>
<ul class="simple">
<li><p>The import program cannot handle CSV files where 2 lines have the same value of row and begin_column - this restriction is similar to that imposed on <a class="reference external" href="https://github.com/GenomicsDB/GenomicsDB/wiki/Importing-VCF-data-into-GenomicsDB#fields-information">loading VCFs</a>.</p></li>
</ul>
</div>
</div>
<div class="section" id="compression-schemas">
<h2>Compression Schemas<a class="headerlink" href="#compression-schemas" title="Permalink to this headline">¶</a></h2>
<p>GenomicsDB supports several compression schemas, which can be configured during build time.</p>
<table class="colwidths-given docutils align-default" id="id5">
<caption><span class="caption-text">Compression Schemas</span><a class="headerlink" href="#id5" title="Permalink to this table">¶</a></caption>
<colgroup>
<col style="width: 25%" />
<col style="width: 25%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Compression Alogirthm</p></th>
<th class="head"><p>Build Variable Value</p></th>
<th class="head"><p>Notes</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>None</p></td>
<td><p>0</p></td>
<td><p>Default value if not specified</p></td>
</tr>
<tr class="row-odd"><td><p>Gzip</p></td>
<td><p>1</p></td>
<td><p>Description</p></td>
</tr>
<tr class="row-even"><td><p>Zstd</p></td>
<td><p>2</p></td>
<td><p>Description</p></td>
</tr>
<tr class="row-odd"><td><p>Lz4</p></td>
<td><p>3</p></td>
<td><p>Description</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="best-practices">
<h2>Best Practices<a class="headerlink" href="#best-practices" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>best practices</p></li>
</ul>
</div>
</div>
<div class="section" id="supported-storage">
<h1>Supported Storage<a class="headerlink" href="#supported-storage" title="Permalink to this headline">¶</a></h1>
<p>GenomicsDB has support for several filesystems.</p>
<div class="section" id="local-cluster">
<h2>Local / Cluster<a class="headerlink" href="#local-cluster" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><em>POSIX</em>: Tested on Centos 6 + 7</p></li>
<li><p><em>HDFS</em>: HDFS support</p></li>
</ul>
</div>
<div class="section" id="cloud">
<h2>Cloud<a class="headerlink" href="#cloud" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><em>AWS</em>: Support for S3 and EMRFS</p></li>
<li><p><em>GCP</em>: Support for G3</p></li>
<li><p><em>Azure</em>: Support for Azure Blob storage</p></li>
</ul>
</div>
</div>
<div class="section" id="examples-tutorials">
<h1>Examples / Tutorials<a class="headerlink" href="#examples-tutorials" title="Permalink to this headline">¶</a></h1>
<div class="section" id="python-notebook">
<h2>Python Notebook<a class="headerlink" href="#python-notebook" title="Permalink to this headline">¶</a></h2>
<p>Example Coming Soon</p>
</div>
<div class="section" id="scala-notebook">
<h2>Scala Notebook<a class="headerlink" href="#scala-notebook" title="Permalink to this headline">¶</a></h2>
<p>Example Coming Soon</p>
</div>
</div>
<div class="section" id="advanced">
<h1>Advanced<a class="headerlink" href="#advanced" title="Permalink to this headline">¶</a></h1>
<div class="section" id="performance-tuning">
<h2>Performance Tuning<a class="headerlink" href="#performance-tuning" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>multi sample vs single sample</p></li>
<li><p>array fragmentation</p></li>
<li><p>tile sizes</p></li>
</ul>
</div>
<div class="section" id="detailed-explanations">
<h2>Detailed Explanations<a class="headerlink" href="#detailed-explanations" title="Permalink to this headline">¶</a></h2>
</div>
</div>
<div class="section" id="using-with-gatk">
<h1>Using with GATK<a class="headerlink" href="#using-with-gatk" title="Permalink to this headline">¶</a></h1>
<p>GenomicsDB is packaged into
<a class="reference external" href="https://software.broadinstitute.org/gatk/documentation/article?id=11091">gatk4</a>
and benefits qualitatively from a large user base.</p>
</div>
<div class="section" id="documentation-work-in-progress">
<h1>Documentation WORK-IN-PROGRESS<a class="headerlink" href="#documentation-work-in-progress" title="Permalink to this headline">¶</a></h1>
<p><a class="reference external" href="https://opensource.org/licenses/MIT"><img alt="License: MIT" src="https://img.shields.io/badge/License-MIT-yellow.svg" /></a></p>
<p>GenomicsDB is a highly performant scalable data storage written in C++ for importing, querying and transforming genomic
variant data.</p>
<div class="section" id="supported-platforms-and-filesystems">
<h2>Supported platforms and filesystems:<a class="headerlink" href="#supported-platforms-and-filesystems" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Linux and MacOS.</p></li>
<li><p>POSIX, HDFS, EMRFS(S3), GCS and Azure Blob.</p></li>
</ul>
<div class="section" id="included-are">
<h3>Included are<a class="headerlink" href="#included-are" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>JVM/Spark wrappers that allow for streaming <a class="reference external" href="https://samtools.github.io/htsjdk/javadoc/htsjdk/htsjdk/variant/variantcontext/VariantContext.html">VariantContext</a> buffers to/from the C++ layer among other functions.</p></li>
</ul>
<p>GenomicsDB jars with native libraries and only zlib dependencies are regularly published on <a class="reference external" href="https://mvnrepository.com/artifact/org.genomicsdb/genomicsdb">Maven Central</a>.
* Native tools for incremental ingestion of variants in the form of VCF/BCF/CSV into GenomicsDB for performance.
* MPI and Spark support for parallel querying of GenomicsDB.</p>
<p>GenomicsDB is packaged into
<a class="reference external" href="https://software.broadinstitute.org/gatk/documentation/article?id=11091">gatk4</a>
and benefits qualitatively from a large user base.</p>
<p>The GenomicsDB documentation for users is hosted as a Github wiki:
<a class="reference external" href="https://github.com/GenomicsDB/GenomicsDB/wiki">https://github.com/GenomicsDB/GenomicsDB/wiki</a></p>
<p>GenomicsDB is open source and all participation is welcome. Please read
the <a class="reference external" href="contrib/README.md">guidelines</a> to help with contributions.</p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, GenomicsDB.org

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>